app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/LectureListTransformer.kt

```diff
@@ -0,0 +1,97 @@
+package nerd.tuxmobil.fahrplan.congress.repositories
+
+import android.support.v4.util.SparseArrayCompat
+import nerd.tuxmobil.fahrplan.congress.models.Lecture
+
+class LectureListTransformer(private val prioritizedRoomProvider: PrioritizedRoomProvider) {
+
+    /**
+     * Transform data loaded from the database to a data structure that we would like to use in
+     * UI code.
+     * Then transform it to the (legacy) data structure that the UI is currently using.
+     */
+    fun legacyTransformLectureList(day: Int, lectures: List<Lecture>): LegacyLectureData {
+        val dayRoomData = transformLectureList(day, lectures)
+
+        // Set Lecture.roomIndex to match the index in DayRoomData.roomDataList
+        dayRoomData.roomDataList.mapIndexed { index, roomData ->
+            roomData.lectures.forEach { it.roomIndex = index }
+        }
+
+        // Map room title to room index
+        val roomsMap = dayRoomData.roomDataList
+                .mapIndexed { index, roomData -> roomData.room to index }
+                .toMap()
+
+        // Map column index to room index
+        // This is simple identity map (0 -> 0, 1 -> 1, etc.)
+        val roomList = SparseArrayCompat<Int>()
+        for (index in 0..dayRoomData.roomDataList.size) {
+            roomList.put(index, index)
+        }
+
+        return LegacyLectureData(
+                lectureListDay = day,
+                lectureList = lectures,
+                roomsMap = roomsMap,
+                roomList = roomList,
+                roomCount = dayRoomData.roomDataList.size,
+                dayRoomData = dayRoomData
+        )
+    }
+
+    fun transformLectureList(day: Int, lectures: List<Lecture>): DayRoomData {
+        // Pre-populate the map with prioritized rooms
+        val roomMap = prioritizedRoomProvider.prioritizedRooms
+                .map { it to mutableListOf<Lecture>() }
+                .toMap()
+                .toMutableMap()
+
+        val sortedLectures = lectures.sortedBy { it.roomIndex }
+        for (lecture in sortedLectures) {
+            val lecturesInRoom = roomMap.getOrPut(lecture.room) { mutableListOf() }
```

Comment:
After sorting lectures `roomIndex` is no longer of interest. By adding the room names and their corresponding lecture to the `roomMap` the former `roomIndex` in each lectures differs from the `roomMap` index in their are located now. This can become very confusing. I therefore would recommend to reset and set the `roomIndex` in each lecture to some value, e.g. `NOT_INITIALIZED`. This hopefully prevents any code further upstream from using the `roomIndex` value.

Reply:
How about marking the field in `Lecture` as `@Deprecated` and removing it eventually? Setting it to a `NOT_INITIALIZED` value doesn't help to avoid using it when writing code. And accessing the value at runtime might not lead to fast crashes, but instead lead to hard to debug problems much further down the line.

Reply:
Nowhere really. Maybe include a note that it will be removed after the UI code is switched from using the contents of `LegacyLectureData` to `DayRoomData`.

Reply:
Do you mean `roomIndex` becomes obsolete for building the room columns as the algorithm will come up its own, independent indices?

Reply:
The roomIndex value in the database can be used to sort the rows when `AppRepository` reads the lecture data. Then neither `LectureListTransformer` not any UI code has to know about it.